system/security/digisign/digisign.cpp

```diff
@@ -0,0 +1,247 @@
+                         char buff[120];                            \
+                         ERR_error_string(ERR_get_error(), buff);   \
+                         throw MakeStringException(-1, str, buff);  \
+                       }
+
+
+class _CDigitalSignatureManager : implements IDigitalSignatureManager
+{
+private:
+    StringAttr   publicKeyFile;
+    StringBuffer publicKeyBuff;
+    StringAttr   privateKeyFile;
+    StringBuffer privateKeyBuff;
+    StringBuffer passphraseBuff;
+    bool         signingConfigured;
+    bool         verifyingConfigured;
+
+public:
+    _CDigitalSignatureManager()
+    {
+        signingConfigured = false;
+        verifyingConfigured = false;
+#ifdef _USE_OPENSSL
+        //query private key file location from environment.conf
+        const char * cert=0, * privKey=0, * passPhrase=0;
+        bool rc = queryHPCCPKIKeyFiles(&cert, &privKey, &passPhrase);
+        if (!isEmptyString(cert))
+            publicKeyFile.set(cert);
+        if (!isEmptyString(privKey))
+            privateKeyFile.set(privKey);
+        if (!isEmptyString(passPhrase))
+            passphraseBuff.set(passPhrase);
+        signingConfigured = !publicKeyFile.isEmpty();
+        verifyingConfigured = !privateKeyFile.isEmpty();
+#else
+        WARNLOG("CDigitalSignatureManager: Platform built without OPENSSL!");
+#endif
+    }
+
+    virtual ~_CDigitalSignatureManager()
+    {
+    }
+
+    bool isDigiSignerConfigured()
+    {
+        return signingConfigured;
+    }
+
+    bool isDigiVerifierConfigured()
+    {
+        return verifyingConfigured;
+    }
+
+    //Create base 64 encoded digital signature of given text string
+    bool digiSign(const char * text, StringBuffer & b64Signature)
+    {
+        if (!signingConfigured)
+            throw MakeStringException(-1, "digiSign:Creating Digital Signatures not configured");
+
+        CriticalBlock b(digiSignCrit);
+
+        if (privateKeyBuff.isEmpty())
+            privateKeyBuff.loadFile(privateKeyFile.str());
+        if (privateKeyBuff.isEmpty())
+            throw MakeStringException(-1, "digiSign:Cannot load private key file");
+
+#ifdef _USE_OPENSSL
+        //create an RSA object from private key
+        BIO * keybio = BIO_new_mem_buf((void*) privateKeyBuff.str(), -1);
+        if (keybio == nullptr)
+        {
+            EVP_THROW("digiSign:BIO_new_mem_buf: %s");
+        }
+
+        RSA * rsa = PEM_read_bio_RSAPrivateKey(keybio, nullptr, nullptr, (void*)passphraseBuff.str());
+        if (rsa == nullptr)
+        {
+            BIO_free_all(keybio);
+            EVP_THROW("digiSign:PEM_read_bio_RSAPrivateKey: %s");
+        }
+        BIO_free_all(keybio);
+
+        //With the RSA object, create the digest and digital signature
+        EVP_MD_CTX * RSASignCtx = EVP_MD_CTX_create(); //allocate, initializes and return a digest context
+        if (nullptr == RSASignCtx)
+        {
+            EVP_THROW("digiSign:EVP_MD_CTX_create returned NULL: %s");
```

Comment:
This will leak rsa if it fails.  May be easiest to move be after the priKey assignment as the verify function does.